cli/src/ui.rs

```diff
@@ -267,43 +266,63 @@ pub enum PaginationChoice {
     Auto,
 }
 
+enum PagerConfig {
+    Disabled,
+    Builtin,
+    Command(CommandNameAndArgs),
+}
+
 impl Ui {
     pub fn with_config(config: &StackedConfig) -> Result<Ui, CommandError> {
         let formatter_factory = prepare_formatter_factory(config, &io::stdout())?;
         Ok(Ui {
             quiet: config.get("ui.quiet")?,
             formatter_factory,
-            pager_cmd: config.get("ui.pager")?,
-            paginate: config.get("ui.paginate")?,
+            pager: Self::pager_config(config)?,
             progress_indicator: config.get("ui.progress-indicator")?,
             output: UiOutput::new_terminal(),
         })
     }
 
     pub fn reset(&mut self, config: &StackedConfig) -> Result<(), CommandError> {
         self.quiet = config.get("ui.quiet")?;
-        self.paginate = config.get("ui.paginate")?;
-        self.pager_cmd = config.get("ui.pager")?;
+        self.pager = Self::pager_config(config)?;
         self.progress_indicator = config.get("ui.progress-indicator")?;
         self.formatter_factory = prepare_formatter_factory(config, &io::stdout())?;
         Ok(())
     }
 
+    fn pager_config(config: &StackedConfig) -> Result<PagerConfig, CommandError> {
+        let PaginationChoice::Auto = config.get("ui.paginate")? else {
+            return Ok(PagerConfig::Disabled);
+        };
+        let pager_cmd: CommandNameAndArgs = config.get("ui.pager")?;
+        match pager_cmd {
+            CommandNameAndArgs::String(name) if name == BUILTIN_PAGER_NAME => {
+                Ok(PagerConfig::Builtin)
+            }
+            _ => Ok(PagerConfig::Command(pager_cmd)),
+        }
+    }
+
     /// Switches the output to use the pager, if allowed.
     #[instrument(skip_all)]
     pub fn request_pager(&mut self) {
-        match self.paginate {
-            PaginationChoice::Never => return,
-            PaginationChoice::Auto => {}
+        if let PagerConfig::Disabled = self.pager {
+            return;
         }
         if !matches!(&self.output, UiOutput::Terminal { stdout, .. } if stdout.is_terminal()) {
             return;
         }
 
-        let use_builtin_pager = matches!(
-            &self.pager_cmd, CommandNameAndArgs::String(name) if name == BUILTIN_PAGER_NAME);
-        let new_output = if use_builtin_pager {
-            UiOutput::new_builtin_paged()
+        let new_output = match &self.pager {
+            PagerConfig::Disabled => {
+                unreachable!(
+                    "earlier check avoids possible syscall when pager disabled, may not be worth \
+                     it"
+                )
```

Comment:
nit: `return` here? It's probably okay to test `is_terminal()` unconditionally.

Reply:
Thanks, I was guessing you might know what's best here.